web-services/security/src/test/java/datawave/security/login/DatawavePrincipalLoginModuleTest.java

```diff
@@ -269,7 +271,179 @@ public void testGetRoleSetsLeavesRequiredRoles() throws Exception {
+
+        replayAll();
+        boolean success = datawaveLoginModule.login();
+        assertFalse("Login did not succeed.", success);
+        assertEquals(userDN, expected.getUserDN());
+
+        verifyAll();
+    }
+
+    @Test(expected = CredentialException.class)
+    public void testDirectRolesFailServer() throws Exception {
+        /**
+         *  Chain is User -> S1 -> S2.  S2 is terminal server.  Verified that s2
+         *  does not have the appropriate authorized role for terminal server (directRole).  This
+         *  will fail the check in the #DatawavePrincipalLoginModule.login()
+         *  This will prevent the chain from accessing any endpoint which requires
+         *  authorized roles
+         */
+
+        String issuerDN = DnUtils.normalizeDN(testServerCert.getIssuerDN().getName());
+        String serverDN = DnUtils.normalizeDN("CN=testServer.example.com, OU=iamnotaperson, OU=acme");
+        SubjectIssuerDNPair server1 = SubjectIssuerDNPair.of(serverDN, issuerDN);
+        String otherServerDN = DnUtils.normalizeDN("CN=otherServer.example.com, OU=iamnotaperson, OU=acme");
+        SubjectIssuerDNPair server2 = SubjectIssuerDNPair.of(otherServerDN, issuerDN);
+        String proxiedSubjects = "<" + userDN.subjectDN() + "><" + otherServerDN + ">";
+        String proxiedIssuers = "<" + userDN.issuerDN() + "><" + issuerDN + ">";
+        DatawaveCredential datawaveCredential = new DatawaveCredential(testServerCert, proxiedSubjects, proxiedIssuers);
+        callbackHandler.name = datawaveCredential.getUserName();
+        callbackHandler.credential = datawaveCredential;
+
+        List<String> userRoles = Arrays.asList("Role1", "AuthorizedUser");
+        List<String> s1Roles = Arrays.asList("Role2", "AuthorizedServer");
+        List<String> s2Roles = Arrays.asList("Role3", "OtherRequiredRole");
+
+        DatawaveUser user = new DatawaveUser(userDN, UserType.USER, null, userRoles, null, System.currentTimeMillis());
+        DatawaveUser s1 = new DatawaveUser(server1, UserType.SERVER, null, s1Roles, null, System.currentTimeMillis());
+        DatawaveUser s2 = new DatawaveUser(server2, UserType.SERVER, null, s2Roles, null, System.currentTimeMillis());
+
+        /**
+         *  s2 has OtherRequiredRole is an auhtorizedRole, but not a directRole
+         *   so this will fail the check in #DatawavePrincipalLoginModule.login()
+         *   so chain User -> S1 -> S2 will fail
+         */
+
+        DatawavePrincipal expected = new DatawavePrincipal(Lists.newArrayList(user, s1, s2));
+
+        expect(securityDomain.getKeyStore()).andReturn(serverKeystore);
+        expect(securityDomain.getTrustStore()).andReturn(truststore);
+        expect(datawaveUserService.lookup(datawaveCredential.getEntities())).andReturn(expected.getProxiedUsers());
+
+        replayAll();
+
+        boolean success = datawaveLoginModule.login();
+        assertFalse("Login did not succeed.", success);
```

Comment:
Same comment as above.  Maybe "login should fail, but succeded" or "login should fail"